tests/unit/lms/models/group_info_test.py

```diff
@@ -43,6 +44,59 @@ def test_application_instance_cant_be_None(self, db_session, group_info):
         ):
             db_session.flush()
 
+    def test_instructors_defaults_to_an_empty_list(self):
+        assert GroupInfo().instructors == []
+
+    def test_upsert_instructor_when_no_existing_instructors(self):
+        group_info = GroupInfo()
+        instructor = factories.HUser()._asdict()
+
+        group_info.upsert_instructor(instructor)
+
+        assert group_info.instructors == [instructor]
+
+    def test_upsert_instructor_when_existing_instructors(self):
+        group_info = GroupInfo()
+        existing_instructors = group_info.instructors = [
+            factories.HUser()._asdict(),
+            factories.HUser()._asdict(),
+            factories.HUser()._asdict(),
+        ]
+        new_instructor = factories.HUser(username="new_instructor")._asdict()
+
+        group_info.upsert_instructor(new_instructor)
+
+        assert group_info.instructors == existing_instructors + [new_instructor]
+
+    def test_upsert_instructor_when_existing_matching_instructor(self):
+        group_info = GroupInfo()
+        updated_instructor = factories.HUser()._asdict()
+        existing_instructors = group_info.instructors = [
+            factories.HUser()._asdict(),
+            factories.HUser(username=updated_instructor["username"])._asdict(),
+            factories.HUser()._asdict(),
+        ]
+
+        group_info.upsert_instructor(updated_instructor)
+
+        assert group_info.instructors == [
+            existing_instructors[0],
+            updated_instructor,
+            existing_instructors[2],
+        ]
+
+    def test_upsert_instructor_when_existing_equal_instructor(self):
+        group_info = GroupInfo()
+        existing_instructors = group_info.instructors = [
+            factories.HUser()._asdict(),
+            factories.HUser()._asdict(),
+            factories.HUser()._asdict(),
+        ]
```

Comment:
I can't remember what it is, but there's a way of doing 3 in a batch isn't there?

Reply:
It's complicated by the fact that I need dicts (I was calling `_asdict()` but:

```diff
diff --git a/tests/unit/lms/models/group_info_test.py b/tests/unit/lms/models/group_info_test.py
index d062cc42..149d49db 100644
--- a/tests/unit/lms/models/group_info_test.py
+++ b/tests/unit/lms/models/group_info_test.py
@@ -1,3 +1,4 @@
+import factory
 import pytest
 from sqlalchemy.exc import IntegrityError
 
@@ -87,11 +88,9 @@ class TestGroupInfo:
 
     def test_upsert_instructor_when_existing_equal_instructor(self):
         group_info = GroupInfo()
-        existing_instructors = group_info.instructors = [
-            factories.HUser()._asdict(),
-            factories.HUser()._asdict(),
-            factories.HUser()._asdict(),
-        ]
+        existing_instructors = group_info.instructors = factory.build_batch(
+            dict, 3, FACTORY_CLASS=factories.HUser
+        )
 
         group_info.upsert_instructor(existing_instructors[1])
```

Or of course you could do something like `[u._asdict() for u in factories.HUser.build_batch(3)]`